### ISSUE TITLE
Loading indicator component throws an axe (accessibility plugin) violation in the newest ruleset

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/vets-design-system-documentation#366]
# Bug Report

- [ ] I’ve searched for any related issues and avoided creating a duplicate issue.

## What happened

The newest version of the [Deque axe browser plugin](https://deque.com/axe) flags our [Loading Indicator](https://design.va.gov/components/loading-indicators) component because it doesn't have an accessible label. This will cause our CI/CD pipeline to break when we [upgrade the axe-core library](https://github.com/department-of-veterans-affairs/va.gov-team/issues/16544). That work is planned for after the holiday. It is blocked by this item and #357.

This issue affects any applications that use the Loading Indicator to let assistive technology know the app / data is being loaded asynchronously.

## What I expected to happen
- [ ] Axe reports zero violations of type "ARIA progressbar nodes must have an accessible name"
- [ ] Existing styles are retained
## Related Issues
* #357 

## Reproducing
- Formation version: Latest
- Device: MacBook Pro, MacOS Catalina
- Browser: Chrome latest

Steps to reproduce:

1. Open https://design.va.gov/components/loading-indicators
1. Open Chrome / Firefox dev tools
1. Run the axe browser plugin (assumes the latest plugin has been added to your browser)
1. Verify the serious violation "ARIA progressbar nodes must have an accessible name" appears in your dev tools

## Potential Code Updates
I took a look at how we might update the markup for our Progress bar. Based on research from the Codepen linked below, I think we should take the existing `aria-valuetext` string and put it into a new `aria-label` attribute. I'll ask @DanielleThierryUSDSVA to take a look and advise what would be good text for the `aria-valuetext`.

For context, VoiceOver and NVDA will read the aria-label, then the aria-valuetext immediately following. JAWS will only read the aria-valuetext.

* https://codepen.io/tpierce_402/pen/KKgNJLN

```diff
<div>
  <div
+   aria-label="Please wait while we load the application for you."
-   aria-valuetext="Please wait while we load the application for you."
+   aria-valuetext=TBD
    class="loading-indicator"
    role="progressbar"
    tabindex="0">
  </div>
  Please wait while we load the application for you.
</div>
```

## Screenshot
![Screen Shot 2021-01-05 at 1 57 25 PM](https://user-images.githubusercontent.com/934879/103693591-28354800-4f5f-11eb-8dfb-c05509b63e77.png)

**Comment by cvalarida** [comment]
# Action items for @1Copenut 
- Determine what the `aria-valuetext` should be

# Action items for @cvalarida
- Determine how the loading indicators are used right now
  - Specifically, whether or not they're all using the `label` prop
  - I'll leave a comment on this ticket with the result of this discovery
- Also make sure we update the loading indicator on the React app landing pages
  - I'll do this once I get the text to put in there from @1Copenut

**Comment by bkjohnson** [comment]
Putting this here as part of the investigation

![image.png](https://images.zenhubusercontent.com/5db75d6f1484010001787ee1/f318a44b-f8cf-4f0a-a4c5-5c414f8f39d9)

**Comment by bkjohnson** [comment]
Here is a list of all the files in `vets-website` that use `<LoadingIndicator>`, according to the following command:

`rg "component-library/LoadingIndicator" -l`

<details>

  <summary>Files</summary>

  ```
src/applications/caregivers/components/SubmitError/DownloadLink.jsx
src/applications/caregivers/containers/App.jsx
src/applications/search/containers/SearchApp.jsx
src/applications/debt-letters/components/DebtLettersWrapper.jsx
src/applications/resources-and-support/components/ResourcesAndSupportSearchApp.jsx
src/applications/facility-locator/components/ResultsList.jsx
src/applications/facility-locator/containers/ProviderDetail.jsx
src/applications/facility-locator/containers/FacilityDetail.jsx
src/applications/claims-status/components/ClaimDetailLayout.jsx
src/applications/validate-mhv-account/containers/ValidateMHVAccount.jsx
src/applications/validate-mhv-account/containers/Main.jsx
src/applications/coronavirus-chatbot/components/ChatbotWrapper.jsx
src/applications/yellow-ribbon/containers/SearchResults/index.jsx
src/applications/static-pages/vre-chapter36/Chapter36CTA.js
src/applications/static-pages/covid-vaccine-updates-cta/widget.jsx
src/applications/static-pages/facilities/FacilityDetailWidget.jsx
src/applications/static-pages/facilities/OtherFacilityListWidget.jsx
src/applications/static-pages/facilities/BasicFacilityListWidget.jsx
src/applications/static-pages/facilities/FacilityListWidget.jsx
src/applications/static-pages/facilities/FacilityPatientSatisfactionScoresWidget.jsx
src/applications/static-pages/facilities/FacilityAppointmentWaitTimesWidget.jsx
src/applications/static-pages/facilities/FacilityMapWidget.jsx
src/applications/static-pages/view-modify-dependents/view-dependents-cta/containers/ViewDependentsCTA.jsx
src/applications/static-pages/health-care-manage-benefits/components/CernerCallToAction/index.js
src/applications/claims-status/containers/YourClaimsPageV2.jsx
src/applications/claims-status/containers/AppealStatusPage.jsx
src/applications/static-pages/view-payment-history/ViewPaymentHistoryCTA.js
src/applications/claims-status/containers/AdditionalEvidencePage.jsx
src/applications/static-pages/view-modify-dependent/686-cta/containers/Form686CTA.jsx
src/applications/static-pages/vre-chapter31/Chapter31CTA.js
src/applications/claims-status/containers/AppealInfo.jsx
src/applications/claims-status/containers/DocumentRequestPage.jsx
src/applications/letters/containers/Main.jsx
src/applications/coronavirus-vaccination/components/Layout.jsx
src/applications/coronavirus-vaccination/components/Form.jsx
src/applications/coronavirus-vaccination/components/SelectFacilityWidget/PatientFacilities.jsx
src/applications/hca/containers/IntroductionPage.jsx
src/applications/vre/28-8832/containers/App.jsx
src/applications/vre/28-1900/components/VeteranInformationViewComponent.jsx
src/applications/gi-sandbox/containers/GiBillApp.jsx
src/applications/find-forms/containers/SearchResults.jsx
src/applications/hca/containers/IDPage.jsx
src/applications/financial-status-report/containers/App.jsx
src/applications/vaos/project-cheetah/components/VAFacilityPage/index.jsx
src/applications/personalization/profile/components/Profile.jsx
src/applications/personalization/profile/components/connected-apps/ConnectedApps.jsx
src/applications/vaos/new-appointment/components/CommunityCareProviderSelectionPage/ProviderSelectionField.jsx
src/applications/vaos/new-appointment/components/ClinicChoicePage.jsx
src/applications/vaos/new-appointment/components/VAFacilityPage/VAFacilityPageV2.jsx
src/applications/vaos/routes.jsx
src/applications/personalization/view-dependents/components/ViewDependentsList/ViewDependentsList.jsx
src/applications/vaos/components/calendar/CalendarWidget.jsx
src/applications/vaos/components/LoadingOverlay.jsx
src/applications/vaos/appointment-list/components/RequestedAppointmentsList.jsx
src/applications/vaos/appointment-list/components/FutureAppointmentsList.jsx
src/applications/vaos/appointment-list/components/FutureAppointmentsListV2.jsx
src/applications/vaos/new-appointment/components/VAFacilityPage/NoValidVAFacilities.jsx
src/applications/vaos/appointment-list/components/PastAppointmentsList/index.jsx
src/applications/vaos/new-appointment/components/VAFacilityPage/index.jsx
src/applications/vaos/appointment-list/components/AppointmentsPage/index.jsx
src/applications/vaos/appointment-list/components/UpcomingAppointmentsList.jsx
src/applications/vaos/express-care/index.jsx
src/applications/personalization/rated-disabilities/components/TotalRatedDisabilities.jsx
src/applications/vaos/components/VAOSApp/index.jsx
src/applications/auth/containers/AuthApp.jsx
src/applications/personalization/dashboard/containers/PrescriptionsWidget.jsx
src/applications/personalization/dashboard/containers/ClaimsAppealsWidget.jsx
src/applications/personalization/dashboard/containers/AppointmentsWidget.jsx
src/applications/personalization/view-representative/components/ViewRepresentativeBody.jsx
src/applications/verify/containers/VerifyApp.jsx
src/applications/coronavirus-vaccination/components/Unsubscribe.jsx
src/applications/personalization/rated-disabilities/components/RatedDisabilityList.jsx
src/applications/personalization/dashboard-2/components/DashboardWrapper.jsx
src/applications/health-care-questionnaire/questionnaire/containers/QuestionnaireWrapper.jsx
src/applications/health-care-questionnaire/questionnaire/containers/App.jsx
src/applications/personalization/preferences/containers/SetPreferences.jsx
src/applications/personalization/preferences/containers/PreferencesWidget.jsx
src/applications/health-care-questionnaire/list/containers/QuestionnaireApp/index.jsx
src/applications/personalization/preferences/containers/JAWSSetPreferences.jsx
src/applications/third-party-app-directory/containers/ThirdPartyAppList/index.jsx
src/applications/gi/containers/VetTecSearchPage.jsx
src/applications/gi/containers/VetTecEstimateYourBenefits.jsx
src/applications/gi/containers/SearchPage.jsx
src/applications/gi/containers/EstimateYourBenefits.jsx
src/applications/gi/containers/GiBillApp.jsx
src/applications/gi/containers/ProfilePage.jsx
src/applications/disability-benefits/2346/containers/App.jsx
src/applications/disability-benefits/all-claims/content/confirmation-poll.jsx
src/applications/disability-benefits/all-claims/containers/AddPerson.jsx
src/applications/disability-benefits/all-claims/containers/ITFWrapper.jsx
src/applications/disability-benefits/686c-674/containers/App.jsx
src/applications/health-care-questionnaire/list/questionnaire-list/components/Home/index.jsx
src/applications/disability-benefits/686c-674/containers/IntroductionPage.jsx
src/applications/disability-benefits/view-payments/components/view-payments-lists/ViewPaymentsLists.jsx
src/applications/disability-benefits/996/containers/IntroductionPage.jsx
src/applications/terms-and-conditions/containers/MhvTermsAndConditions.jsx
src/applications/post-911-gib-status/containers/ServiceAvailabilityBanner.jsx
src/platform/site-wide/cta-widget/index.js
src/platform/forms/save-in-progress/SaveInProgressIntro.jsx
src/platform/forms/save-in-progress/ApplicationStatus.jsx
src/platform/forms/save-in-progress/RoutedSavableApp.jsx
src/platform/user/authorization/containers/MHVApp.jsx
src/platform/user/authorization/containers/RequiredTermsAcceptanceView.jsx
src/platform/user/authorization/components/RequiredLoginView.jsx
src/applications/post-911-gib-status/containers/Main.jsx
src/applications/edu-benefits/feedback-tool/components/SchoolSelectField.jsx
src/platform/monitoring/DowntimeNotification/containers/DowntimeNotification.jsx
src/platform/utilities/ui/asyncLoader.jsx
```

</details>

**Comment by bkjohnson** [comment]
@1Copenut, do you want the `aria-valuetext` to be different than the message that gets displayed?

**Comment by 1Copenut** [comment]
@bkjohnson
I talked for a good bit this morning with @cvalarida about this loading indicator. I've been noodling on the VAOS project's Loading Indicator that has no message, but have had a hard time spotting it in the actual app. With that said, I think we can simplify the component a bit. 

```javascript
return (
  <div className="loading-indicator-container">
    <div aria-hidden="true" className="loading-indicator" />
    <span
      ref={(msg) => (this.spinnerMsg = msg)}
      tabIndex="-1"
    >
      {message}
    </span>
  </div>
);
```

This change would benefit from three additions to the documentation:

1. Teams should be using the Loading Indicator with a message. It's required in the propTypes object.
2. We should encourage teams to use the `setFocus` prop when they are loading a page for the first time. This will move focus to the `span[tabIndex="-1"]` so screen readers will announce the message.
3. Teams will need to manage focus after the Loading Indicator drops out of the page. Some apps like the VAOS Express Care already do this. I tested the new Express Care request route, and verified the loading indicator took focus correctly, then the H1 on the page when it was ready.

I believe this change should be announced broadly when the PR goes up, so we're not inadvertently causing breaks upstream.

This fix also handles the axe error because we're removing the role from the "spinner" and hiding it from AT in favor of the message.

**Comment by bkjohnson** [comment]
The Design System team has discussed this, and due to the growing scope we have decided to pause implementation and hand it off to @JoeTice.

**Comment by 1Copenut** [comment]
@ddzz I had an interesting conversation with Noah and Jeff on VAOS about this one. We're trying out an experiment on the VAOS project where we change the role of this component to "status". Noah is going to test with screen readers. Shouldn't affect scope, but will inform my comments. Hope to have some results late today or Monday morning.

**Comment by 1Copenut** [comment]
This issue has been resolved with [PR #28](https://github.com/department-of-veterans-affairs/component-library/pull/28) in Darius' comment above. Moving to close.
